testsuite/ClosedWellTyped.hs

```diff
@@ -0,0 +1,150 @@
+module ClosedWellTyped (
+    ClosedWellTyped(..)
+    ) where
+
+import Control.Applicative as A ((<$>))
+import Control.Monad.State.Lazy
+import Data.Text.Lazy (Text, pack)
+import Test.QuickCheck
+
+import Morte.Core
+
+
+newtype ClosedWellTyped = ClosedWellTyped { unClosedWellTyped :: Expr X }
+    deriving Show
+
+data Env a = Env {
+      bindings :: [(Var, Expr a)]
+    , uniques  :: [Text]
+    }
+
+type GenExpr e a   = StateT (Env e) Gen a
+type GenExprPair e = StateT (Env e) Gen (Expr e,Expr e)
+
+instance Arbitrary ClosedWellTyped where
+    arbitrary = sized rndExpr
+
+
+rndExpr :: Int -> Gen ClosedWellTyped
+rndExpr n = ClosedWellTyped <$> evalStateT genExpr (initEnv n)
+
+initEnv :: Int -> Env a
+initEnv n = Env [] $ map (pack . show) [1..n]
+
+extend :: Var -> Expr e -> GenExpr e ()
+extend x t = modify (\env -> env { bindings = (x,t) : bindings env })
+
+select :: [(Int,GenExprPair e,Env e -> Bool)] -> GenExprPair e
```

Comment:
I don't quite understand what you are getting at here and how this could help. The only partial function here is ``frequency`` (ok, and by extension every function that uses select), but if the preconditions are right, ``frequency`` will never fail, we can always generate *something*(it would be nice to express this but I think it's overkill with regard to the size of the implementation). The partiality of select helped me sometimes when I got the preconditions wrong.
This is the second iteration of the generator, the first was a backtracking one and the use of ``Maybe`` permeated through every function and made de code a little harder to grasp.

I'll give your suggestion a try, maybe I misunderstood the implications.